File: Week 25/cpp/2.rabin_karp.cpp
/*
 * WEEK 25 - C++ ADVANCED DSA
 * Topic: Rabin-Karp String Matching (rolling hash)
 * File: 2.rabin_karp.cpp
 *
 * CONCEPT:
 *     Slide a fixed-size window of T and compare its rolling-hash fingerprint
 *     to the hash of P. On a hash collision verify literally. Expected
 *     runtime is O(n + m).
 *
 * KEY POINTS:
 *     - Polynomial rolling hash mod a large prime, BASE = 256.
 *     - h = BASE^(m-1) mod PRIME used to subtract the outgoing character.
 *     - Always verify a candidate window literally — hashes can collide.
 *     - Worst case O(n*m) under adversarial input.
 *
 * ALGORITHM / APPROACH:
 *     h = BASE^(m-1) mod PRIME
 *     compute p_hash and the first t_hash
 *     for i in 0..n-m:
 *         if p_hash == t_hash and T[i..i+m] == P: report i
 *         t_hash = (BASE * (t_hash - T[i]*h) + T[i+m]) mod PRIME
 *
 * C++-SPECIFIC NOTES vs JAVA:
 *     - Use long long to stay clear of overflow before mod.
 *     - text.substr(i, m) materialises a copy on each candidate; for hot
 *       paths replace with a hand-written compare loop.
 *     - char promotes to int naturally when arithmetic is performed.
 *
 * DRY RUN:
 *     P = "AAA", T = "AAAAAA". h = 256^2 mod PRIME = 65536.
 *     p_hash = ((65*256 + 65)*256 + 65) = 4_276_545 (within mod).
 *     Each window shares the same fingerprint -> matches [0,1,2,3].
 *
 * COMPLEXITY:
 *     Time  : O(n + m) expected, O(n*m) worst.
 *     Space : O(1) extra.
 *
 * Compile: g++ -std=c++17 -O2 2.rabin_karp.cpp -o rk
 */

#include <iostream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::string;
using std::vector;

vector<int> search(const string& text, const string& pattern) {
    const long long BASE  = 256;
    const long long PRIME = 1000000007LL;

    vector<int> results;
    int n = static_cast<int>(text.size());
    int m = static_cast<int>(pattern.size());
    if (m == 0 || m > n) return results;

    long long h = 1;
    for (int i = 0; i < m - 1; i++) h = (h * BASE) % PRIME;

    long long p_hash = 0, t_hash = 0;
    for (int i = 0; i < m; i++) {
        p_hash = (BASE * p_hash + static_cast<unsigned char>(pattern[i])) % PRIME;
        t_hash = (BASE * t_hash + static_cast<unsigned char>(text[i]))    % PRIME;
    }

    for (int i = 0; i <= n - m; i++) {
        if (p_hash == t_hash) {
            if (text.compare(i, m, pattern) == 0) {
                results.push_back(i);
            }
        }
        if (i < n - m) {
            t_hash = (BASE * (t_hash - static_cast<unsigned char>(text[i]) * h)
                      + static_cast<unsigned char>(text[i + m])) % PRIME;
            if (t_hash < 0) t_hash += PRIME;
        }
    }
    return results;
}

static void print_vec(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        cout << v[i];
        if (i + 1 < v.size()) cout << ", ";
    }
    cout << "]";
}

int main() {
    cout << "=== Rabin-Karp Pattern Matching (C++) ===" << endl;
    string text = "ABABDABACDABABCABAB";
    string pattern = "ABABCABAB";
    cout << "Text   : " << text << endl;
    cout << "Pattern: " << pattern << endl;
    cout << "Matches: "; print_vec(search(text, pattern)); cout << endl;   // [10]

    cout << "\n--- Overlapping matches ---" << endl;
    string t2 = "AAAAAA", p2 = "AAA";
    cout << "Text   : " << t2 << endl;
    cout << "Pattern: " << p2 << endl;
    cout << "Matches: "; print_vec(search(t2, p2)); cout << endl;          // [0,1,2,3]
    return 0;
}

/*
 * NOTES (vs Java baseline):
 *     - C++ promotes char to int implicitly; we cast to unsigned char first
 *       so non-ASCII bytes (>= 0x80) don't become negative in the hash.
 *     - text.compare(i, m, pattern) avoids materialising a temporary string
 *       that Java's substring/equals chain would in the textbook version.
 */

File: Week 25/rust/2.rabin_karp.rs
// WEEK 25 - RUST ADVANCED DSA
// Topic: Rabin-Karp String Matching (rolling hash)
// File: 2.rabin_karp.rs
//
// CONCEPT:
//     Slide a fixed-size window of T and compare its rolling-hash fingerprint
//     to the hash of P. On a hash collision verify literally. Expected
//     runtime is O(n + m).
//
// KEY POINTS:
//     - Polynomial rolling hash mod a 1e9+7 prime; BASE = 256.
//     - h = BASE^(m-1) mod PRIME used to subtract the outgoing character.
//     - Use u64 so the multiplication BASE * hash never overflows.
//     - Because u64 underflows wrap around, add PRIME explicitly before any
//       subtraction in the update step.
//
// ALGORITHM / APPROACH:
//     h = BASE^(m-1) mod PRIME
//     compute p_hash and the first t_hash
//     for i in 0..=n-m:
//         if p_hash == t_hash and T[i..i+m] == P: report i
//         t_hash = (BASE * ((t_hash + PRIME) - (T[i]*h) % PRIME) + T[i+m]) mod PRIME
//
// RUST-SPECIFIC NOTES vs JAVA:
//     - No signed `long`; unsigned arithmetic requires explicit underflow
//       protection by adding PRIME before subtraction.
//     - Operate on &[u8] for clean indexing.
//     - vec![] returned when m == 0 || m > n.
//
// DRY RUN:
//     P = "AAA", T = "AAAAAA". Each window shares the same fingerprint, so
//     all four positions [0,1,2,3] are reported after a literal check.
//
// COMPLEXITY:
//     Time  : O(n + m) expected, O(n*m) worst.
//     Space : O(1) extra.

fn search(text: &str, pattern: &str) -> Vec<usize> {
    const BASE: u64 = 256;
    const PRIME: u64 = 1_000_000_007;

    let t = text.as_bytes();
    let p = pattern.as_bytes();
    let (n, m) = (t.len(), p.len());
    if m == 0 || m > n {
        return vec![];
    }

    let mut h: u64 = 1;
    for _ in 0..m - 1 {
        h = (h * BASE) % PRIME;
    }

    let mut p_hash: u64 = 0;
    let mut t_hash: u64 = 0;
    for i in 0..m {
        p_hash = (BASE * p_hash + p[i] as u64) % PRIME;
        t_hash = (BASE * t_hash + t[i] as u64) % PRIME;
    }

    let mut results = Vec::new();
    for i in 0..=n - m {
        if p_hash == t_hash && &t[i..i + m] == p {
            results.push(i);
        }
        if i < n - m {
            // Add PRIME before subtraction to avoid unsigned underflow.
            t_hash = (BASE * ((t_hash + PRIME) - (t[i] as u64 * h) % PRIME)
                      + t[i + m] as u64) % PRIME;
        }
    }
    results
}

fn main() {
    println!("=== Rabin-Karp Pattern Matching (Rust) ===");
    let text = "ABABDABACDABABCABAB";
    let pattern = "ABABCABAB";
    println!("Text   : {}", text);
    println!("Pattern: {}", pattern);
    println!("Matches: {:?}", search(text, pattern));    // [10]

    println!("\n--- Overlapping matches ---");
    let t2 = "AAAAAA";
    let p2 = "AAA";
    println!("Text   : {}", t2);
    println!("Pattern: {}", p2);
    println!("Matches: {:?}", search(t2, p2));           // [0, 1, 2, 3]
}

// NOTES (vs Java baseline):
//     - Java's signed `long` allows negative intermediate values that we
//       correct by adding PRIME. Rust's u64 wraps on underflow, so we add
//       PRIME *before* subtracting.
//     - For mission-critical use add a second (BASE, PRIME) pair (double
//       hashing) to drive collision probability to ~1/PRIME^2.
